### PR TITLE
[Config] Deprecate TreeBuilder::root

### PR DESCRIPTION
**Pull request by gharlan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | 
| Fixed tickets | #29876
| License       | MIT
| Doc PR        | —

Alternative idea to #31015. Or is the `root` method still needed?

It would look like this:

Screenshot 2019-04-09 01 15 04

### CODE REVIEW

File: src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php [hunk 1]
```diff
@@ -43,9 +44,13 @@ public function __construct(string $name = null, string $type = 'array', NodeBui
      * @return ArrayNodeDefinition|NodeDefinition The root node (as an ArrayNodeDefinition when the type is 'array')
      *
      * @throws \RuntimeException When the node type is not supported
+     *
+     * @deprecated since Symfony 4.3, pass the root name to the constructor instead
      */
     public function root($name, $type = 'array', NodeBuilder $builder = null)
     {
+        @trigger_error(sprintf('The "%s()" method (called for "%s") is deprecated since Symfony 4.3, pass the root name to the constructor instead.', __METHOD__, $name), E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: maybe `called for the "%s" configuration` (something that hints what %s is)?

Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: And remove the brackets I think.

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php [hunk 2]
```diff
@@ -43,9 +44,13 @@ public function __construct(string $name = null, string $type = 'array', NodeBui
      * @return ArrayNodeDefinition|NodeDefinition The root node (as an ArrayNodeDefinition when the type is 'array')
      *
      * @throws \RuntimeException When the node type is not supported
+     *
+     * @deprecated since Symfony 4.3, pass the root name to the constructor instead
      */
     public function root($name, $type = 'array', NodeBuilder $builder = null)
     {
+        @trigger_error(sprintf('The "%s()" method called for the "%s" configuration is deprecated since Symfony 4.3, pass the root name to the constructor instead.', __METHOD__, $name), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: idea: we may want to tweak the deprecation message of the constructor in the `4.2` in the same way to mention the configuration

**Review by xabbuh — CHANGES_REQUESTED** [review]
Oh, please mention the deprecation in the component changelog file as well as in the upgrade files for 4.3 and 5.0.

File: src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php (hunk 2 above)
Comment (gharlan) [reply src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: When the constructor is called without the root name, we can not mention the not given name in the deprecation message ;)

Comment (xabbuh) [reply src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: 🙈

**Comment by gharlan** [comment]
Status: Needs Review

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @gharlan.

**Comment by sstok** [comment]
This makes compatibility for older versions much harder as you still need to call `root()` in those cases, while passing the name to the constructor was was already possible.

```php
$treeBuilder = new TreeBuilder('rollerworks_password_strength');
$rootNode = $treeBuilder->root('rollerworks_password_strength'); // BC for Symfony < 4.2
```

What about checking if a name was already set in the constructor, and it equals the passed name. 
In this case, return `$this->root` and do nothing, otherwise throw an deprecation instead.

To ensure forward compatibility you still need to check if the `root()` method exists. But least you don't get any deprecation warnings.

```php
$treeBuilder = new TreeBuilder('rollerworks_password_strength');

if (method_exists($treeBuilder, 'root')) {
    $rootNode = $treeBuilder->root('rollerworks_password_strength'); // BC for Symfony < 4.2
} else {
    $rootNode = $treeBuilder->getRootNode();
}
```

**Comment by gharlan** [comment]
As far as I know many libs already use `method_exists`.

**Comment by sstok** [comment]
Ah I completely missed #27476 😛 

https://github.com/symfony/symfony/pull/27476#issuecomment-457910730 gives a good suggestion 
```php
$treeBuilder = new TreeBuilder("my_node");
$rootNode = method_exists($treeBuilder, "getRootNode")
    ? $treeBuilder->getRootNode()
    : $treeBuilder->root("my_node");
```

👍

**Pull request merged**